src/hmem_synapseai.c

```diff
@@ -43,24 +43,44 @@
 #include "habanalabs/synapse_api.h"
 
 struct synapseai_ops {
+	synStatus (*synInitialize)(void);
+	synStatus (*synDestroy)(void);
+	synStatus (*synDeviceGetCount)(uint32_t* pCount);
 	int (*hcclLookupDMABuff)(uint64_t addr, uint64_t size, int* fd);
 };
 
 static void *synapseai_handle;
 static struct synapseai_ops synapseai_ops;
 
-/**
- * @brief Initialize SynapseAI hmem interface by dlopen the SynapseAI library
- * 
- * @return On success, return 0. On failure, return negative error code.
- */
-int synapseai_init(void)
+int synapseai_dl_init()
 {
+	int err = FI_SUCCESS;
 	synapseai_handle = dlopen("libSynapse.so", RTLD_NOW);
 	if (!synapseai_handle) {
 		FI_INFO(&core_prov, FI_LOG_CORE,
 			"Failed to dlopen libSynapse.so\n");
-		return -FI_ENOSYS;
```

Comment:
this doesn't look right, you shouldn't dlclose in this case as the dlopen failed.